# Module 02 Homework - Pandas
Using a dataset from _Wine Spectator_, a wine magazine, we will practice the material that we covered in notebook M02_Colab04 to M02_Colab06. This notebook will cover data transformation, grouping, and sorting using pandas.

Created by Scott Du (016955484)  
Last updated: 2/22/2026

In [53]:
csvurl="https://gist.githubusercontent.com/clairehq/" + \
        "79acab35be50eaf1c383948ed3fd1129/raw/407a02139ae1e134992b90b4b2b8c329b3d73a6a/winemag-data-130k-v2.csv"
import pandas as pd
wine = pd.read_csv(csvurl)

**Data cleaning**  
Notice that the first column is redundant. Part of data analysis is cleaning and removing redundancy. How would you drop the redundant column inplace, that is overwrite the dataframe.

In [54]:
wine.drop('Unnamed: 0', axis='columns', inplace=True)
wine.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

#### Question 1: ####  
What is the mean of the points column?

In [55]:
# Your code here . . .
wine.describe()

points         price
count  65499.000000  60829.000000
mean      88.434037     35.232932
std        3.030310     39.477858
min       80.000000      4.000000
25%       86.000000     17.000000
50%       88.000000     25.000000
75%       91.000000     42.000000
max      100.000000   2500.000000

The mean of the points column is 88.434037.

#### Question 2: ####  
How many countries are present in this dataset? (Only count each country once)

In [56]:
# Your code here . . .
countries = wine['country'].unique()
print(len(countries))
countries

42


array(['Italy', 'Portugal', 'US', 'Spain', 'France', 'Germany',
       'Argentina', 'Chile', 'Australia', 'Austria', 'South Africa',
       'New Zealand', 'Israel', 'Hungary', 'Greece', 'Romania', 'Mexico',
       'Canada', nan, 'Turkey', 'Czech Republic', 'Slovenia',
       'Luxembourg', 'Croatia', 'Georgia', 'Uruguay', 'England',
       'Lebanon', 'Serbia', 'Brazil', 'Moldova', 'Morocco', 'Peru',
       'India', 'Bulgaria', 'Cyprus', 'Armenia', 'Switzerland',
       'Bosnia and Herzegovina', 'Ukraine', 'Slovakia', 'Macedonia'],
      dtype=object)

There are 42 different countries present in this dataset.

['Italy', 'Portugal', 'US', 'Spain', 'France', 'Germany',
       'Argentina', 'Chile', 'Australia', 'Austria', 'South Africa',
       'New Zealand', 'Israel', 'Hungary', 'Greece', 'Romania', 'Mexico',
       'Canada', nan, 'Turkey', 'Czech Republic', 'Slovenia',
       'Luxembourg', 'Croatia', 'Georgia', 'Uruguay', 'England',
       'Lebanon', 'Serbia', 'Brazil', 'Moldova', 'Morocco', 'Peru',
       'India', 'Bulgaria', 'Cyprus', 'Armenia', 'Switzerland',
       'Bosnia and Herzegovina', 'Ukraine', 'Slovakia', 'Macedonia']

#### Question 3: ####
How many times does each country appeared in this dataset? Show each country and the corresponding count (show counts in ascending order)

In [57]:
# Your code here . . .
wine['country'].value_counts(ascending=True)

country
Bosnia and Herzegovina        1
Slovakia                      1
Armenia                       1
Luxembourg                    4
India                         4
Switzerland                   4
Ukraine                       5
Czech Republic                6
Macedonia                     6
Cyprus                        6
Serbia                        7
Peru                          8
Morocco                      11
Lebanon                      20
Moldova                      30
Mexico                       31
Brazil                       31
England                      36
Georgia                      37
Slovenia                     39
Turkey                       43
Croatia                      44
Hungary                      61
Uruguay                      61
Romania                      67
Bulgaria                     68
Canada                      108
Greece                      242
Israel                      259
New Zealand                 733
South Africa                737
Germany                    1051
Australia                  1177
Austria                    1635
Argentina                  1907
Chile                      2258
Portugal                   2963
Spain                      3409
Italy                     10005
France                    11174
US                        27177
Name: count, dtype: int64

#### Question 4: ####
Create a variable `adjusted_price` containing the adjusted price which is the price subtracted by the average price. *This is called **"centering" transformation** - a method commonly used in the preprocessing step before applying various machine learning algorithms.*

In [58]:
# Your code here . . .
avg_price = wine['price'].mean()
adjusted_price = wine['price'] - avg_price
adjusted_price

0              NaN
1       -20.232932
2       -21.232932
3       -22.232932
4        29.767068
           ...    
65494     9.767068
65495   -13.232932
65496   -15.232932
65497    -4.232932
65498   -25.232932
Name: price, Length: 65499, dtype: float64

#### Question 5: ####
What is the title of the wine that has the highest points-to-price ratio in the dataset?

In [59]:
# Your code here . . .
wine['points-price-ratio'] = wine['points'] / wine['price']
max = wine['points-price-ratio'].idxmax()
wine.iloc[max]['title']

'Bandit NV Merlot (California)'

#### Question 6: ####
Create a series `flavor_counts` that contains two values: the number of wines that has the word "tart" in the `description` column and the number of wines that has the word "berries" in the `description` column. The index of the Series should be "Tart" and "Berries" for the corresponding values.

In [60]:
# Look for the word "tart", not case sensitive, not a substring of another word.
# Same applies to the word "berries"
# Your code here . . .
lowercase_wine = wine['description'].str.lower()
wine_by_words = lowercase_wine.str.split(' ')

tarts = wine_by_words.apply(lambda x : 'tart' in x).value_counts()[True]
berries = wine_by_words.apply(lambda x : 'berries' in x).value_counts()[True]

flavor_counts = pd.Series([tarts, berries],
                          index=['Tarts', 'Berries'],
                          name='Count')
flavor_counts


Tarts      2378
Berries     620
Name: Count, dtype: int64

#### Question 7: ####
Let's convert the points into simple star ratings. A score of 90 or higher counts as 3 stars, a score of at least 80 but less than 90 is 2 stars. Any other score is 1 star.

Also, any wines from France should automatically get 3 stars, regardless of points.

Add this new column `star_ratings` to the dataframe with the number of stars for each wine in the dataset.

In [61]:
# Your code here . . .
wine['star_ratings'] = wine['points'].apply(lambda points : 
    '3 stars' if points >= 90
    else ('2 stars' if points >= 80
          else '1 star'))

wine.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  points-price-ratio star_ratings  
0     White Blend              Nicosia                 NaN      2 stars  
1  Portuguese Red  Quinta dos Avidagos            5.800000      2 stars  
2      Pinot Gris            Rainstorm            6.214286      2 stars  
3        Riesling           St. Julian            6.692308      2 stars  
4      Pinot Noir         Sweet Cheeks            1.338462      2 stars

#### Question 8: ####
Who are the most common wine reviewers in the dataset? Create a Series whose index is the taster_twitter_handle category from the dataset, and whose values count how many reviews each person wrote.

In [62]:
# Your code here . . .
common_wine_reviewers = wine.groupby('taster_twitter_handle').size()
common_wine_reviewers.rename('Reviews', inplace=True)
common_wine_reviewers.sort_values(ascending=False, inplace=True)
common_wine_reviewers

taster_twitter_handle
@vossroger          13045
@wineschach          7752
@kerinokeefe         5313
@paulgwine           4851
@vboone              4696
@mattkettmann        3035
@JoeCz               2605
@wawinereport        2358
@gordone_cellars     2032
@AnneInVino          1769
@laurbuzz             938
@suskostrzewa         593
@worldwineguys        465
@bkfiona               11
@winewchristina         4
Name: Reviews, dtype: int64

#### Question 9: ####
What combination of countries and varieties are most common? Create a Series whose index is a MultiIndexof {country, variety} pairs. For example, a pinot noir produced in the US should map to {"US", "Pinot Noir"}. Sort the values in the Series in descending order based on wine count.

In [63]:
# Your code here . . .
common_combos = wine.groupby(['country', 'variety']).size().sort_values(ascending=False)
common_combos.rename('Count', inplace=True)
common_combos

country    variety                 
US         Pinot Noir                  4918
           Cabernet Sauvignon          3649
           Chardonnay                  3412
France     Bordeaux-style Red Blend    2380
Italy      Red Blend                   1870
                                       ... 
Romania    Rosé                           1
US         Ugni Blanc                     1
           Touriga                        1
           Torrontés                      1
Argentina  Merlot-Cabernet Franc          1
Name: Count, Length: 1304, dtype: int64

#### Question 10 #####
Create a Series whose index is reviewers and whose values is the average score given out by that reviewer.  
*Hint:* You will need the `taster_name` and `points` columns.

In [64]:
# Your code here . . .
avg_score_taster = wine.groupby('taster_name')['points'].mean()
avg_score_taster.rename('Average Score', inplace=True)
avg_score_taster

taster_name
Alexander Peartree    86.014286
Anna Lee C. Iijima    88.380506
Anne Krebiehl MW      90.587903
Carrie Dykes          86.644444
Christina Pickard     89.500000
Fiona Adams           87.090909
Jeff Jenssen          88.273504
Jim Gordon            88.604331
Joe Czerwinski        88.519770
Kerin O’Keefe         88.827969
Lauren Buzzeo         87.831557
Matt Kettmann         90.021087
Michael Schachner     86.904541
Mike DeSimone         89.030303
Paul Gregutt          89.095032
Roger Voss            88.678957
Sean P. Sullivan      88.666243
Susan Kostrzewa       86.408094
Virginie Boone        89.229557
Name: Average Score, dtype: float64